### PR TITLE
Simplify service section of circuit propose wizard

### PR DESCRIPTION
**Pull request by peterschwarz**
This PR reorganizes the circuit form modules and replaces the `ServiceCard` with a combination of a `ServiceTable` and `ServiceForm`.  These provide a more standard, clear way of creating services for a service.

To test:

```
docker-compose -f docker/docker-compose.yaml up --build
``` 

Navigate to [localhost:3031](http://localhost:3031) and create a user, if needed.

Add the "Alice" key via the profile sapling and activate.  The key can be obtained via:

```
docker-compose -f docker/docker-compose.yaml run generate-registry bash
cat /registry/alice.priv
cat /registry/alice.pub
```

Propose a circuit, select both nodes and click next.  You should see the following:
Screen Shot 2020-08-13 at 11 48 09 AM
Click the add service button and you will see:
Screen Shot 2020-08-13 at 11 48 18 AM
After a service is added, it will appear in the table:
Screen Shot 2020-08-13 at 11 49 51 AM

### CODE REVIEW

**Comment by peterschwarz** [comment]
Note, the overlay style/sizing, as well as the table whitespace, will be addressed in a future PR

**Review by dnewh — DISMISSED** [review]

File: saplings/circuits/src/components/forms/propose_circuit/service.js [hunk 1]
```diff
@@ -17,11 +17,11 @@
...
 import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
-import { Service, generateID } from '../data/circuits';
-import { Node } from '../data/nodeRegistry';
-import { PlusButton, MinusButton } from './PlusMinusButton';
+import { Service, generateID } from 'App/data/circuits';
+import { Node } from 'App/data/nodeRegistry';
+import { PlusButton, MinusButton } from 'App/components/PlusMinusButton';
 
-import './ServiceCard.scss';
+import './service.scss';
```
Comment (dplumb94) [inline saplings/circuits/src/components/forms/propose_circuit/service.js:37]: The NodeDisplay and ServiceTable components should probably be split up into two separate files. That way we wouldn't have this case where the scss file and service.js file doesnt follow the recommended naming conventions and we won't have a super big file that exports two different components. Maybe that doesn't belong in this PR though.

File: saplings/circuits/src/components/Chips.js [hunk 2]
```diff
@@ -28,6 +28,7 @@ export function Chips({ nodes, localNodeID, removeFn }) {
         const isLocal = node.identity === localNodeID;
         return (
           <Chip
+            key={node.identity}
```
Comment (dplumb94) [inline saplings/circuits/src/components/Chips.js:31]: We should get this check added as a step in CI if possible

File: saplings/circuits/package.json [hunk 3]
```diff
@@ -41,6 +41,7 @@
     "prop-types": "^15.7.2",
     "react": "^16.13.1",
     "react-dom": "^16.13.1",
+    "react-dropdown": "^1.7.0",
```
Comment (dplumb94) [inline saplings/circuits/package.json:44]: Did you find this component easy to work with? My experience with pre-built dropdown components is that it can be extremely challenging to get things styled properly and consistently (though it was a different dependency). Its also not too hard to build a custom dropdown. 

I agree with not using the default dropdown though.

File: saplings/circuits/src/components/forms/controls/index.js [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+  ])
+};
+
+InputRow.defaultProps = {
+  className: '',
+  children: []
+};
+
+export const Button = ({ label, type, ...props }) => {
+  // this circumvents a linting issue around type
```
Comment (dplumb94) [inline saplings/circuits/src/components/forms/controls/index.js]: What is the lint issue?

File: saplings/circuits/src/components/forms/controls/index.js [hunk 5]
```diff
@@ -0,0 +1,189 @@
+/**
```
Comment (dplumb94) [inline saplings/circuits/src/components/forms/controls/index.js:1]: Any reason not to split these up into their own files?

File: saplings/circuits/src/components/forms/propose_circuit/service.js [hunk 6]
```diff
@@ -14,565 +14,442 @@
...
+      <div className="form-content">
+        <InputRow>
+          {basicInput('service-type', 'Service Type')}
+          {basicInput(
+            'service-id',
+            'Service ID',
+            <Button
+              label={<FontAwesomeIcon icon={faSync} />}
+              className="icon-button"
+              onClick={generateId}
```
Comment (dplumb94) [inline saplings/circuits/src/components/forms/propose_circuit/service.js:404]: Is there ever a case where we actually want to fill out the Service ID manually? It feels a bit awkward to click a button and randomly generate an ID that has no meaning to the user.

File: saplings/circuits/src/components/forms/propose_circuit/service.js [hunk 7]
```diff
@@ -14,565 +14,442 @@
...
-              className="service-input"
-              value={serviceState.service.serviceType}
-              onChange={e => {
-                serviceDispatcher({
-                  type: 'set-service-type',
-                  serviceType: e.target.value
-                });
-              }}
-              disabled={!editMode}
+    <div className="form service-form">
```
Comment (dplumb94) [inline saplings/circuits/src/components/forms/propose_circuit/service.js:393]: For some reason the dropdown border has a different shade from the other inputs
Screen Shot 2020-08-13 at 5 45 27 PM

File: saplings/circuits/src/components/forms/propose_circuit/service.js [hunk 8]
```diff
@@ -14,565 +14,442 @@
...
+  isEdit,
+  onServiceEdit,
+  onServiceDelete
+}) => {
+  let tableBody;
+  if (services.length === 0) {
+    tableBody = (
+      <tr>
+        <td className="no-services" colSpan="5">
+          <center>At least one service must be defined for a circuit.</center>
```
Comment (dplumb94) [inline saplings/circuits/src/components/forms/propose_circuit/service.js:78]: I think this placeholder text shouldn't be selectable. Also I think it might be best to avoid the ultra-low contrast text for placeholder messages for accessibility (though I am aware that this is common...)

File: saplings/circuits/src/components/forms/propose_circuit/service.js [hunk 9]
```diff
@@ -14,565 +14,442 @@
...
+    <div className="form service-form">
+      <span className="service-form-header">Add a Service</span>
+      <div className="form-content">
+        <InputRow>
+          {basicInput('service-type', 'Service Type')}
+          {basicInput(
+            'service-id',
+            'Service ID',
+            <Button
+              label={<FontAwesomeIcon icon={faSync} />}
```
Comment (dplumb94) [inline saplings/circuits/src/components/forms/propose_circuit/service.js:402]: I feel like the look of the faSync icon doesn't really match the plus minus icons on the key value pairs. Also I think they have a different alignment from the fields. (probably not an issue for this PR)

File: saplings/circuits/src/components/forms/propose_circuit/service.js [hunk 10]
```diff
@@ -14,565 +14,442 @@
...
+                content: <div className="no-node-selected">Select a node</div>
+              },
+              ...nodes.map(node => ({
+                value: node.identity,
+                content: <NodeDisplay node={node} />
+              }))
+            ]}
+          />
+        </InputRow>
+        {values['service-args'].map((_, idx) => serviceArgRow(idx))}
```
Comment (dplumb94) [inline saplings/circuits/src/components/forms/propose_circuit/service.js:430]: It isn't really clear in the UI what these key value pairs are for

File: saplings/circuits/src/components/forms/propose_circuit/service.js [hunk 11]
```diff
@@ -14,565 +14,442 @@
...
+
+        <div className="button-group">
```
Comment (dplumb94) [inline saplings/circuits/src/components/forms/propose_circuit/service.js:432]: I think these buttons should be placed outside of the form so that they dont scroll out of view when you add too many key value pairs. 

In general, I think there should be some kind of clear delineation between the form header and the body/content so it doesn't look like the content randomly cuts off at the edge of the container. (For example, the Zenhub issue creation modal's header)

It might make more sense to just have a container for the key value pairs so that only that portion of the form needs to scroll.

File: saplings/circuits/src/components/forms/propose_circuit/service.js [hunk 12]
```diff
@@ -14,565 +14,442 @@
...
-        </div>
-        <div className="service-field bg-color-grey">
-          <div className="field-name">Service arguments</div>
-          <div className="field-input service-arguments">
-            {serviceArguments()}
-          </div>
+          )}
+        </InputRow>
+        <InputRow>
+          <ListBoxSelect
```
Comment (dplumb94) [inline saplings/circuits/src/components/forms/propose_circuit/service.js:409]: I think that the text box of this should align with the edge of the other text boxes, rather than the buttons

**Review by dplumb94 — CHANGES_REQUESTED** [review]
Feel free to ignore any of the stylistic comments if they're just copy pasted and not relevant to the goal of the PR.

File: saplings/circuits/src/components/Chips.js (hunk 2 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/Chips.js:31]: It would be nice, but these errors are emitted at runtime, when built in development mode.  Not sure how to get them at build time.

File: saplings/circuits/src/components/forms/propose_circuit/service.js (hunk 1 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/propose_circuit/service.js:37]: NodeDisplay should probably get moved out (it can probably be used in more places), but the ServiceTable is part of the service module.

Part of this re-layout is to provide better organization for where shared components should reside.  The current organization is slapdash.  As part of an upcoming style/minor functional tweaks will probably include more reorganization.

File: saplings/circuits/package.json (hunk 3 above)
Comment (peterschwarz) [reply saplings/circuits/package.json:44]: I did a reasonably exhaustive search, and this library was one of the only ones I found that allowed you to add classnames to the main parts of its structure.  If you look where its used, I've added custom classes and added styles to the ones that needed.

File: saplings/circuits/src/components/forms/controls/index.js (hunk 4 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/controls/index.js]: It does not like the use of the spread operator with JSX.  Seems like an arbitrary decision, since the use of the operator is extremely valuable for creating custom components that are still highly configurable.

File: saplings/circuits/src/components/forms/controls/index.js (hunk 5 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/controls/index.js:1]: They all belong to this module - I'm not sure I agree with the java-like "one-class-per file" naming convention (that wasn't really being followed in the first place).

File: saplings/circuits/src/components/forms/propose_circuit/service.js (hunk 6 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/propose_circuit/service.js:404]: That was following the previous implementation.  Users may still like the flexibility of using their own 4 letter conventions.

File: saplings/circuits/src/components/forms/propose_circuit/service.js (hunk 9 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/propose_circuit/service.js:402]: Yes, they are a bit strange - the plus-minus buttons are weirdly sized.  I'll see if I can fix it.

File: saplings/circuits/src/components/forms/propose_circuit/service.js (hunk 11 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/propose_circuit/service.js:432]: Can you supply some image so I know what you mean?

File: saplings/circuits/src/components/forms/propose_circuit/service.js (hunk 10 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/propose_circuit/service.js:430]: Would some place holder text and a section header help?

File: saplings/circuits/src/components/forms/propose_circuit/service.js (hunk 12 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/propose_circuit/service.js:409]: I agree, I think it would be more clear if the "generate id" button was part of the control, instead of next to it.  I'll see what I can do here.

File: saplings/circuits/src/components/forms/propose_circuit/service.js (hunk 8 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/propose_circuit/service.js:78]: I used the same style code as the placeholder text for the circuit table.  Perhaps that can be part of the future style-tweak PR.

File: saplings/circuits/src/components/forms/propose_circuit/service.js (hunk 7 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/propose_circuit/service.js:393]: What browser is that in?  This is what I see in firefox:
Screen Shot 2020-08-14 at 9 50 59 AM

File: saplings/circuits/src/components/forms/propose_circuit/service.js (hunk 1 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/propose_circuit/service.js:37]: Actually, it's a private type - it's reused in the service form and the service table.

File: saplings/circuits/src/components/forms/controls/index.js (hunk 4 above)
Comment (dnewh) [reply saplings/circuits/src/components/forms/controls/index.js]: We disabled this lint in other recent projects

File: saplings/circuits/src/components/forms/propose_circuit/service.js (hunk 7 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/propose_circuit/service.js:393]: Nevermind, found this in chrome

File: saplings/circuits/src/components/forms/propose_circuit/service.js (hunk 9 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/propose_circuit/service.js:402]: Updated.

File: saplings/circuits/src/components/forms/propose_circuit/service.js (hunk 10 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/propose_circuit/service.js:430]: Added a section header

File: saplings/circuits/src/components/forms/propose_circuit/service.js (hunk 11 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/propose_circuit/service.js:432]: I've re-styled this a bit better: the form is centered, with some padding, and this overflows better.

File: saplings/circuits/src/components/forms/propose_circuit/service.js (hunk 12 above)
Comment (peterschwarz) [reply saplings/circuits/src/components/forms/propose_circuit/service.js:409]: I've moved the button into the field, so it is more connected to the control.

File: saplings/circuits/src/components/forms/controls/index.js (hunk 4 above)
Comment (dplumb94) [reply saplings/circuits/src/components/forms/controls/index.js]: I would prefer if we disabled the lint. I think a good approach for this would be to take the airbnb lint config (as it provides a good base), edit as we prefer, and then package it up. Then we can just pull in that commonly agreed upon config to every relevant project.

**Review by dplumb94 — CHANGES_REQUESTED** [review]
I would recommend that we disable the unnecessary lint rule for now. We should come up with a common lint config and package it up to be imported by each project.

**Comment by peterschwarz** [comment]
Let’s save that for another PR. It’s a separate task from this one.

File: saplings/circuits/src/components/forms/controls/index.js (hunk 4 above)
Comment (dplumb94) [reply saplings/circuits/src/components/forms/controls/index.js]: We can do this in another PR

**Review by dplumb94 — APPROVED** [review]

**Review by dnewh — APPROVED** [review]
